src/cockpit/samples.py

```diff
@@ -192,49 +192,55 @@ class CGroupSampler(Sampler):
-                if sys.maxsize > value > 0:
-                    samples[name][cgroup] = value / 1000
-
-    def read_cgroup_integer_stat(self, samples, path, cgroup, name):
+    @staticmethod
+    def read_cgroup_integer_stat(rootfd, statfile, include_zero=False, key=b''):
         # Not every stat is available, such as cpu.weight
         try:
-            with open(path) as stat:
-                # Some samples such as "memory.max" contains "max" when there is a no limit
-                try:
-                    value = int(stat.read().strip())
-                except ValueError:
-                    return
-
-                if sys.maxsize > value > 0:
-                    samples[name][cgroup] = value
-        except FileNotFoundError:
-            pass
+            fd = os.open(statfile, os.O_RDONLY, dir_fd=rootfd)
+        except Exception:
+            return
+
+        data = os.read(fd, 1024)
+        os.close(fd)
+
+        if key:
+            start = data.index(key) + len(key)
+            end = data.index(b'\n', start)
+            # cpu.stat are in usecs
+            value = int(data[start:end]) / 1000
+        else:
+            # Some samples such as "memory.max" contains "max" when there is a no limit
+            try:
+                value = int(data)
+            except ValueError:
+                return
+
+        if value > sys.maxsize:
+            return
+
+        if value > 0:
+            return value
+        elif include_zero:
+            return value
 
     def sample(self, samples):
         # TODO: Cgroups v1 support
         if self.cgroups_v2 is None:
             self.cgroups_v2 = os.path.exists('/sys/fs/cgroup/cgroup.controllers')
 
         if self.cgroups_v2_path:
-            for path, _, _ in os.walk(self.cgroups_v2_path):
-                cgroup = path.replace(self.cgroups_v2_path, '')  # TODO: Pathlib?
+            for path, _, _, rootfd in os.fwalk(self.cgroups_v2_path):
+                cgroup = path.replace(self.cgroups_v2_path, '')
 
                 if not cgroup:
                     continue
 
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.current'), cgroup, 'cgroup.memory.usage')
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.max'), cgroup, 'cgroup.memory.limit')
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.swap.current'), cgroup, 'cgroup.memory.sw-usage')
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'memory.swap.max'), cgroup, 'cgroup.memory.sw-limit')
-                self.read_cgroup_integer_stat(samples, os.path.join(path, 'cpu.weight'), cgroup, 'cgroup.cpu.shares')
-                self.read_cgroup_keyed_stat(samples, os.path.join(path, 'cpu.stat'), cgroup, 'cgroup.cpu.usage', 'usage_usec')
+                samples['cgroup.memory.usage'][cgroup] = self.read_cgroup_integer_stat(rootfd, 'memory.current', True)
```

Comment:
This reads a whole lot nicer.  But what do you do with all of those `None` values?

Reply:
Hmm ideally the channel would return False, however everything now works by coincidence more or less...

We request:
```
      { name: "cgroup.cpu.usage", derive: "rate" },
      { name: "cgroup.memory.usage" },
```

As it's an `instanced` metric and a `rate` we will need an `old_value` and `value`. As None is not True the metrics won't be send.